Several tag related improvements

TheBlob42 opened the pull request:
This PR improves tag related functionality with the following:
- list tags as document symbols (#261)
- list tags as workspace symbols
- allow text filtering for auto completion of tags

Obviously these changes are somewhat opinionated but I am open for discussion. Having tags available as symbols makes it easier to find related content. Especially for people using a Zettelkasten or other note taking system with marksman. Quickly finding all notes related to a certain tag or quickly jumping to a relevant section in a long markdown document should ease these workflows.

In regards to the filtering I always wondered why I could not enter any more text after the initial `#` to filter for a specific tag. I found the related comment in the code, but I don't understand why this decision was made.

### Problem to separate headings and tags

Since all symbols in markdown are of the type "String" (nothing else really fits) they are hard to distinguish from another. For the workspace symbols I have made the prefix (e.g. "Tag", "H1", "H2" etc.) filterable as well so that one could easily filter only for tags or heading. However this only makes sense for people using some sort of fuzzy searching instead of a literal string matching where this could lead to usability issues. I am open for other ideas on how to solve this

Once this is clarified I would also check the formatting and add tests for these new cases :slightly_smiling_face:

artempyanykh commented on Marksman/Compl.fs:
```diff
@@ -639,8 +639,6 @@ module Completions =
             Some
                 { CompletionItem.Create(label) with
                     Detail = Some detail
-                    // Use input as filter text to avoid any extra filtering on the editor's side
```
@TheBlob42 did this line cause problems?

TheBlob42 replied:
Exactly. The `FilterText` property is the string that the user input is matched against when filtering the completion candidates (see [here](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem)). If you set this value for example to `Some "foo"` and trigger the completion then typing "foo" in your editor should still show all completion candidates (this is an extreme example I know).

The problem here was that the filter text was set to the user input at the time of the completion candidates creation. Therefore every "additional" character would not match this text anymore and the completion candidates would be gone.

artempyanykh replied:
Thanks for the fix here! I think there's a _same-ish_ problem with other types of completions... I haven't fully figured out the deal with client side filtering of completion candidates.

TheBlob42 commented:
The one thing missing currently are dedicated tests for the DocumentSymbold and WorkspaceSymbol functionalities. If you're cool with the changes and the layout I would take some time to add those 🙂

TheBlob42 commented:
I have cleaned up my changes and the commit history. I have also added some missing tests (workspace symbols) and adapted the existing ones to incorporate the changes in relation to tags, so that they get tested as well.

@artempyanykh please have a look. I am happy to discuss any changes being made :slightly_smiling_face:

artempyanykh reviewed (approved):
Looks great! Appreciate the improvement! 💖

artempyanykh commented:
I'll push a new release shortly

petobens commented:
Hi @TheBlob42 and @artempyanykh  thanks for the change! Works great (feel free to close #261). 
One minior comment: if I have an url like `https://docs.google.com/spreadsheets/d/1Rr7GcfjUwhZfylUk0FGL0Et4kYF5etl49OpRwRmhhlg/edit#gid=1116260529&range=D12` marksman will treat this as tag. Is it possible to filter out these cases?

TheBlob42 commented:
@petobens good catch :thinking: would probably be worth its own issue

petobens commented:
Sure! Done. Thanks!

Gelio commented:
I'm super excited this was added! I've been using tags for some time and I missed the LSP features for them.

One enhancement I can suggest is supporting [nested tags from Obsidian](https://help.obsidian.md/Editing+and+formatting/Tags#Nested%20tags), like `#home/chore`, `#work/meeting`. This would make my Neovim authoring experience even better

artempyanykh commented:
@Gelio please do create a separate issue for this. Would be great if you could elaborate more on what kind of support you want from nested tags, example use cases.

Gelio commented:
Sure thing, I created https://github.com/artempyanykh/marksman/issues/267
